Drag multiple items in tree-view

UziTech opened the pull request:
### Description of the Change

* Allow multiple items to be dragged to a folder. This is a rebase of #793 

![multi-drag screenshot](https://cloud.githubusercontent.com/assets/6645121/14222364/8df00814-f835-11e5-80b2-1184c6d454a1.gif)

* Allow multi-select to work like it is supposed to

![shift + ctrl + click screenshot](https://user-images.githubusercontent.com/97994/30246499-7440e930-95c1-11e7-9941-9e0a6de4301f.gif)


### Benefits

Multi-select will work like other file explorers

### Possible Drawbacks

none

### Applicable Issues

Fixes #232
Supersedes and closes https://github.com/atom/tree-view/pull/793

50Wliu commented:
@UziTech what is the purpose of 0f7a4b7?

UziTech commented:
@50Wliu

When trying to drag multiple files you click and hold on one of the selected files and drag to a folder.

When selecting only a single selected file when multiple are selected you just click on the selected file.

We need to make sure the user is not dragging before deselecting all files except for the one that is clicked on.  In order to do that we need to do the deselect on mouseup. 

We still want most mouse operations to occur on mousedown. Only when we need to make sure there is no dragging do we do the operation on mouseup.

Ben3eeE commented:
I'm signing up for testing here since I tested the original PR.

andyfleming commented:
Great work! Let's get this done!

I can't believe we can't drag multiple files in atom yet 😆 .

UziTech commented:
@Ben3eeE anything holding this up?

Ben3eeE commented:
@UziTech Sorry been swamped with other things. I will take a look as soon as I get a chance to, hopefully I will have some time this weekend ⚡️

UziTech commented:
I rebased again and fixed the merge conflicts in the tests

UziTech commented:
@Ben3eeE Any way we can get this out the door so I don't have to keep rebasing?

arthurlacoste commented:
Please, merge this feature !

arthurlacoste commented:
Please, merge this feature !

arthurlacoste commented:
Please, merge this feature !

UziTech commented:
Windows tests are failing because of #1203 

@50Wliu @Ben3eeE any chance this can get merged

UziTech commented:
@Ben3eeE is this waiting on anything?

UziTech commented:
@50Wliu @Ben3eeE @nathansobo any reason this isn't getting merged?

nathansobo commented:
@ungb @rsese Can one of you test this?

50Wliu commented:
@Ben3eeE and I have been trying to find a time where we can test this together and do a code review.

Ben3eeE commented:
@UziTech Sorry for taking so long before I could get to this.

Here is a test plan with some scenarios I can think of. If anyone can think of anything else then please let me know :bowing_man: 

## Test plan

- [ ] Selecting and deselecting multiple files and folders with ctrl/shift click (cmd on macOS?`)
  - [ ] Windows
  - [ ] Linux
  - [ ] macOS
- [ ] Multiple drag and drop
  - [ ] Multiple files
  - [ ] Multiple folders
  - [ ] Multiple files and folders
- [ ] Initiating drag 
  - [ ] From a file
  - [ ] From a folder
    - [ ] From a collapsed folder
    - [ ] From an expanded folder
    - [ ] From an expanded folder when you have files in the folder selected /cc: https://github.com/atom/tree-view/pull/793#issuecomment-222503727
- [ ] Dropping in another Atom window
  - [ ] Multiple files
  - [ ] Multiple folders
  - [ ] Multiple files and folders
- [ ] Drag image should look correct
- [ ] Conflicts
  - [ ] Dragging multiple files/folders when the project root folder is included in the selection /cc: https://github.com/atom/tree-view/pull/525
    - [ ] Dropping in another Atom window when the root directory is selected
  - [ ] Dropping in other locations of the Atom window

UziTech commented:
I fixed a few of the highlighting issues when dragging that occurred because we use the `'selected'` class to highlight folder that are dragged over.

I feel like the right way to fix it would be to use a different class (e.g. `'drag-over-directory'`) to highlight the directory but that would require a change in all themes.

Not that it would be a common bug but if someone checked `treeView.getSelectedEntries()` when dragging files over a directory, the directory would be added to the selected entries which would not be correct.

UziTech commented:
That actually doesn't fix the highlighting issue when dragging from a different Atom window. The only way I can think of to fix it would be to use a different class.

UziTech commented:
I believe I fixed the issue where a folder would get deselected when dragging over it with files from another Atom window.

I added the class `drag-over` as well as `selected` so themes still highlight the folder but only remove the `selected` class on leave if it also has the `drag-over` class

@Ben3eeE I manually tested all of those cases on a Windows 10 and they seem to work as expected. One other case I came across that doesn't work but probably should is dropping a selection on a file. I would expect that the selection would be moved into the same folder as the file that is being dropped on, but right now the move operation is just canceled when dropping on a file.

Actual:
![screenshot2](https://user-images.githubusercontent.com/97994/35518014-d70dee5e-04d5-11e8-98d0-3579d58dea5f.gif)

Expected:
![screenshot](https://user-images.githubusercontent.com/97994/35518016-db91c2f2-04d5-11e8-9bc5-00f2103a0bbe.gif)

Ben3eeE commented:
> One other case I came across that doesn't work but probably should is dropping a selection on a file. I would expect that the selection would be moved into the same folder as the file that is being dropped on, but right now the move operation is just canceled when dropping on a file.

Tested this quick and it appears to also be an issue on `master` so it's not going to block this PR from moving forward. Do you want to create an issue for this?

UziTech commented:
I just added it to this pr

50Wliu commented on lib/tree-view.coffee:
```diff
@@ -869,7 +873,7 @@ class TreeView
 
     entryName = path.basename(initialPath)
-    newPath = "#{newDirectoryPath}/#{entryName}".replace(/\s+$/, '')
+    newPath = "#{newDirectoryPath}#{path.sep}#{entryName}".replace(/\s+$/, '')
```
Nit: could you remove this change and add it in a separate PR?  It should also be `path.join(newDirectoryPath, entryName)`.

I ask because I'm also fixing this same line in both #1180 and #1173, so it would probably be better to just submit this change in a different PR.

50Wliu commented on lib/tree-view.coffee:
```diff
@@ -894,28 +898,71 @@ class TreeView
-        return
+    e.stopPropagation()
+
+    # TODO: meta+click and ctrl+click should not do the same thing on Windows.
```
Is fixing this something that would require a significant rework of the specs?

UziTech replied:
Yes. Any spec that simulates a right click does so by setting e.metaKey to true.

It is also not that big of a deal because in Windows meta(Windows key)+click isn't something people usually do.

50Wliu commented on lib/tree-view.coffee:
```diff
@@ -894,28 +898,71 @@ class TreeView
-        @selectMultipleEntries(entryToSelect)
+    # return early if clicking on a selected entry
+    if entryToSelect.classList.contains('selected')
 
```
:fire: blank line

50Wliu commented on lib/tree-view.coffee:
```diff
@@ -38,6 +38,8 @@ class TreeView
     @roots = []
     @selectedPath = null
+    @selectOnMouseUp = null
+    @lastFocusedElement = null
```
Seems like this makes more sense as `@lastFocusedEntry`?

50Wliu commented on lib/tree-view.coffee:
```diff
@@ -957,103 +1005,132 @@ class TreeView
     @list.classList.remove('full-menu')
     @list.classList.add('multi-select')
 
+  toggleMultiSelectMenu: ->
```
`toggle` generally implies "if on, turn off (and vice versa)".  Therefore a better name for this method might be `showMultiSelectMenuIfNecessary`.

50Wliu commented on lib/tree-view.coffee:
```diff
@@ -957,103 +1005,132 @@ class TreeView
+          newElement.style.paddingRight = "1em"
+          dragImage.append(newElement)
+
+
```
:fire: extra empty line

50Wliu commented:
Going back to your comment here: https://github.com/atom/tree-view/pull/1179#issuecomment-328366447

Perhaps `click` could be used to combine the two cases?

UziTech commented:
I could use `click` but I would still need a variable letting the `onClick` function know if it should execute otherwise it would deselect a file after selecting it on `mousedown`. And we do want to use the `mousedown` function for selecting because someone might want to ctrl+click then just start dragging.

![screenshot](https://user-images.githubusercontent.com/97994/35539360-da5592b0-0516-11e8-848c-f631dade1c59.gif)

adamreisnz commented:
Would be great to have this 👍

50Wliu commented:
Code changes look good to me.  @Ben3eeE if you're good with this as well I think it can :ship:?

Ben3eeE commented:
@UziTech All good from me when testing on Windows 👍 @ungb is going to go over the test plan on macOS and Linux and merge if all is good.

ungb commented:
I'm looking into this now. going off test plan in https://github.com/atom/tree-view/pull/1179#issuecomment-360445015. I'll merge this once I'm done

ungb commented:
I ran into the following while dragging between two atom windows.

See gif for repro.

![error](https://user-images.githubusercontent.com/5624870/35898993-acaabb88-0b7e-11e8-80ee-cca26e05442d.gif)


Repro steps:
Run the following command to create the folders and file
```
mkdir test1 test2
cd test1 && touch a b c && cd ../test2 && touch a b c && cd ..
atom test1 && atom test2
```

1. Drag file `a` from folder `test1` to folder `test2` in the other atom window


**Atom**: 1.23.3 x64
**Electron**: 1.6.15
**OS**: Mac OS X 10.12.6
**Thrown From**: [tree-view](https://github.com/atom/tree-view) package 0.222.0


### Stack Trace

Uncaught TypeError: Cannot read property 'collapse' of null

```
At /Users/ungb/github/packages/tree-view/lib/tree-view.coffee:1126

TypeError: Cannot read property 'collapse' of null
    at TreeView.module.exports.TreeView.onDrop (/packages/tree-view/lib/tree-view.coffee:1126:9)
    at /packages/tree-view/lib/tree-view.coffee:206:47)
```

UziTech commented:
Nice catch

ungb commented:
`Ctrl+click` doesn't work on linux.. it seems to just open up the context menu.  Shift works fine.  

@Ben3eeE does this work on windows?  I can spend some time testing windows today as well.

After the latest change, everything works great on mac!

Ben3eeE commented:
>@Ben3eeE does this work on windows?

@ungb It did when I tested. I have not tested on the most recent commit however.

ungb commented:
@Ben3eeE I think my issues is because I'm running on a VM on mac and it's doing something weird.  I'll take a look when I get home on my windows machine and then try to see if I spin up a VM on there for linux.  So far everything looks good, but every time I hit ctrl+click it opens up the context menu.  I'm noticing this on my VM and happens on the desktop on windows too.

Ben3eeE commented:
@ungb I have Ubuntu installed so I can test this on Linux if you run into more issues because of testing in a VM.

ungb commented:
I've tested this and everything looks good to me.  I'm going to merge this.  I'm going to follow up with the team later on the appveyor failure so we can get this into master.

UziTech commented:
![hallelujah-gif-15](https://user-images.githubusercontent.com/97994/36059690-92172078-0e02-11e8-8e54-7553ba88ff72.gif)

jonathan-dejong commented:
Any news on this? 
From what I can tell with Atom 1.26.1 it is not merged yet.

Ben3eeE commented:
@jonathan-dejong This is merged but it's not released yet because the tests are failing on `atom/atom` when upgrading the tree-view. You may want to subscribe to https://github.com/atom/atom/pull/17088 for updates on when it will be available in a released version of Atom.
